README.md
# pusher-platform-ruby

Pusher Platform SDK for Ruby.

## Installation

Add `pusher-platform` to your Gemfile:

```
gem 'pusher-platform', '~> 0.11.2'
```

## Usage

In order to access Pusher Platform, first instantiate an Instance object:

```ruby
require 'pusher-platform'

pusher = PusherPlatform::Instance.new(
  locator: 'your:instance:locator',
  key: 'key-id:key-secret',
  service_name: 'chatkit',
  service_version: 'v1'
)
```

### Authentication (Rails + Devise)

Instance objects provide an `authenticate` method, which can be used in Rails
controllers to build authentication endpoints. Authentication endpoints issue
access tokens used by Pusher Platform clients to access the API.

Make sure you authenticate the user before issuing access tokens, e.g. by using
the `authenticate_user!` Devise filter.

```ruby
class AuthController < ActionController::Base
  before_action :authenticate_user!

  def auth
    auth_data = pusher.authenticate_with_request(
      request,
      { user_id: current_user.id }
    )
    response.headers = response.headers.merge(auth_data.headers)
    render json: auth_data.body, status: auth_data.status
  end
end
```

### Request API

Instance objects provide a low-level request API, which can be used to contact
Pusher Platform.

```ruby
begin
  response = pusher.request(
    method: "POST",
    path: "feeds/playground",
    headers: {
      "Content-Type": "application/json",
    },
    body: { items: ["test"] }.to_json
  )
  p response.status
  p response.headers
  p response.body
rescue PusherPlatform::ErrorResponse => e
  p e.to_s
rescue PusherPlatform::Error => e
  p e
end
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased](https://github.com/pusher/pusher-platform-ruby/compare/0.11.2...HEAD)

## [0.11.2](https://github.com/pusher/pusher-platform-ruby/compare/0.11.1...0.11.2) - 2018-10-30

### Changes

- Remove rack as a required dependency

## [0.11.1](https://github.com/pusher/pusher-platform-ruby/compare/0.11.0...0.11.1) - 2018-09-18

### Changes

- Makes rack an explicit required dependency in the Gemspec

## [0.11.0](https://github.com/pusher/pusher-platform-ruby/compare/0.10.0...0.11.0) - 2018-09-18

### Changes

- *Breaking* An instance of `SDKInfo` is now required when instantiating an instance of `BaseClient`. This can either be directly provided to the `BaseClient` and then the base client provided to an instance of `Instance` in the `client` key or the `options` object that `Instance`s initializer takes, or you can provide the `SDKInfo` instance to the `Instance` initializer in the `sdk_info` key of its `options` parameter

## [0.10.0](https://github.com/pusher/pusher-platform-ruby/compare/0.9.0...0.10.0) - 2018-09-17

### Changes

- All errors now inherit from `PusherPlatform::Error`
- An exception is raised if no `key` key is present in the hash passed to the `initialize` function for `Instance`

## [0.9.0](https://github.com/pusher/pusher-platform-ruby/compare/0.8.2...0.9.0) - 2018-09-13

### Changes

- Removed defaulting to generating a JWT with the `su: true` claim if no JWT is provided to a call to `request`
- Bump jwt dependency to 2.1.0
- Bump excon dependency to 0.62.0
- JWT algorithm now enforced to be `HS256` when decoding a refresh token

## [0.8.2](https://github.com/pusher/pusher-platform-ruby/compare/0.8.1...0.8.2) - 2018-07-20

### Fixes

- Fixes `ErrorResponse` being incorrectly instantiated in `BaseClient` [#19](https://github.com/pusher/pusher-platform-ruby/issues/19) [@CarlLooksUp](https://github.com/CarlLooksUp)

## [0.8.1](https://github.com/pusher/pusher-platform-ruby/compare/0.8.0...0.8.1) - 2018-05-24

### Fixes

- Fixes `ErrorResponse` being incorrectly instantiated [#18](https://github.com/pusher/pusher-platform-ruby/issues/18)

### Additions

- Adds support for service-specific claims in tokens
- Adds an example of publishing to a feed in [Feeds](https://pusher.com/feeds)

### Changes

- Removed runtime dependency on Rack

## [0.8.0](https://github.com/pusher/pusher-platform-ruby/compare/0.7.0...0.8.0) - 2018-04-20

### Additions

- `authenticate_with_refresh_token` has been added if you want to support the `refresh_token` grant type and return refresh tokens as part of the authentication process
- `authenticate_with_refresh_token_and_request` has been added if you want to support the `refresh_token` grant type and return refresh tokens as part of the authentication process, as well as providing the request object when authenticating

### Changes

- `authenticate` no longer returns a `refresh_token` and no longer accepts the `refresh_token` grant type
- Calls to `authenticate` and `authenticate_with_refresh_token` always return an `AuthenticationResponse` that looks like this (in JSON form):

```ruby
{
  status: 200,
  headers: { "Some-Header" => "some value" },
  body: {
    access_token: "an_access_token",
    token_type: "bearer",
    expires_in: 86400
  }
}
```

where:

* `status` is the suggested HTTP response status code,
* `headers` are the suggested response `headers`,
* `body` holds either the token payload or an appropriate error payload.

Here is an example of the expected usage (using Sinatra), simplified for brevity:

```ruby
post '/' do
  auth_payload = instance.authenticate_with_request(request, { user_id: 'USER_ID' })
  [auth_payload.status, auth_payload.body.to_json]
end
```

## [0.7.0](https://github.com/pusher/pusher-platform-ruby/compare/0.6.0...0.7.0) - 2018-04-12

### Changes

- Rename `authenticate` to `authenticate_with_request`
- `authenticate` (now `authenticate_with_request`) no longer returns a status code. It will either return a `Hash` containing the required authentication payload or an error with information about the reason for authentication failure
- Add some examples

### Additions

- Examples for `authenticate` and `authenticate_with_request`

### Fixes

- Changed module name from `Pusher` to `PusherPlatform` (fixes #15)
- Fixes some `require`s

## [0.6.0](https://github.com/pusher/pusher-platform-ruby/compare/0.5.1...0.6.0) - 2018-01-26

### Changes

- `app` claim in JWTs changed to `instance`

_.. prehistory_


lib/authenticator.rb
require 'jwt'
require 'rack'

module Pusher
  TOKEN_LEEWAY = 30
  TOKEN_EXPIRY = 24*60*60

  class Authenticator
    def initialize(app_id, app_key_id, app_key_secret)
      @app_id = app_id
      @app_key_id = app_key_id
      @app_key_secret = app_id
    end

    # Takes a Rack request to the authorization endpoint and and handles it
    # either returning a new access/refresh token pair, or an error.
    #
    # @param request [Rack::Request] the request to authenticate
    # @return the response object
    def authenticate(request, options)
      form_data = Rack::Utils.parse_nested_query request.body.read
      grant_type = form_data['grant_type']

      if grant_type == "client_credentials"
        return authenticate_with_client_credentials(options)
      elsif grant_type == "refresh_token"
        old_refresh_jwt = form_data['refresh_token']
        return authenticate_with_refresh_token(old_refresh_jwt, options)
      else
        return response(401, {
          error: "unsupported_grant_type"
        })
      end
    end

    private

    def authenticate_with_client_credentials(options)
      return respond_with_new_token_pair(options)
    end

    def authenticate_with_refresh_token(old_refresh_jwt, options)
      old_refresh_token = begin
        JWT.decode(old_refresh_jwt, @app_key_secret, true, {
          iss: "keys/#{@app_key_id}",
          verify_iss: true,
          leeway: 30,
        }).first
      rescue => e
        error_description = if e.is_a?(JWT::InvalidIssuerError)
          "refresh token issuer is invalid"
        elsif e.is_a?(JWT::ImmatureSignature)
          "refresh token is not valid yet"
        elsif e.is_a?(JWT::ExpiredSignature)
          "refresh tokan has expired"
        else
          "refresh token is invalid"
        end

        return response(401, {
          error: "invalid_grant",
          error_description: error_description,
          # TODO error_uri
        })
      end

      if old_refresh_token["refresh"] != true
        return response(401, {
          error: "invalid_grant",
          error_description: "refresh token does not have a refresh claim",
          # TODO error_uri
        })
      end

      if options[:user_id] != old_refresh_token["sub"]
        return response(401, {
          error: "invalid_grant",
          error_description: "refresh token has an invalid user id",
          # TODO error_uri
        })
      end

      return respond_with_new_token_pair(options)
    end

    # Creates a payload dictionary made out of access and refresh token pair and TTL for the access token.
    #
    # @param user_id [String] optional id of the user, ignore for anonymous users
    # @return [Hash] Payload as a hash
    def respond_with_new_token_pair(options)
      access_token = generate_access_token(options)
      refresh_token = generate_refresh_token(options)
      return response(200, {
        access_token: access_token,
        token_type: "bearer",
        expires_in: TOKEN_EXPIRY,
        refresh_token: refresh_token,
      })
    end

    def generate_access_token(options)
      now = Time.now.utc.to_i

      claims = {
        app: @app_id,
        iss: "keys/#{@app_key_id}",
        iat: now - TOKEN_LEEWAY,
        exp: now + TOKEN_EXPIRY + TOKEN_LEEWAY,
        sub: options[:user_id],
      }

      JWT.encode(claims, @app_key_secret, "HS256")
    end

    def generate_refresh_token(options)
      now = Time.now.utc.to_i

      claims = {
        app: @app_id,
        iss: "keys/#{@app_key_id}",
        iat: now - TOKEN_LEEWAY,
        refresh: true,
        sub: options[:user_id],
      }

      JWT.encode(claims, @app_key_secret, "HS256")
    end

    def response(status, body)
      return {
        status: status,
        json: body,
      }
    end
  end
end
